At Long Last, Towels

GoldenAlpharex opened the pull request:
## About The Pull Request
Yeah, you read the title correctly. Towels. At long last, we will have towels, something you'd think would be so basic everyone would have them, considering how you tend to use at least one every day, most of the time.

I could've just ripped them from oldbase, but being myself, I wanted it to be coded well, and to have some solid sprites to match.

#### Tell me more...
So, towels. You can use them in-hand to shape them into something you could wear over your head (head slot), over your chest and waist/legs (suit slot) or over only your waist/legs (suit slot).

When you drop the towel on the floor (sadly applies to putting them down on a table too), if they're not neatly folded, they're going to just collapse into a bit of a pile, which you have to straighten up with ctrl-click (can be done by both humanoids and cyborgs, and can be done at all stages except when a towel is already folded, provided that you're not actively wearing it). You can still hand it over to someone else just fine, but expect it to turn into a mess if you put it in your backpack!

Towels can be used to dry someone up (by clicking on them) when they're wet, which is still to this day a very underused mechanic, which I'm quite likely to open a PR upstream to tie into showers and just generally speaking, water's effect when being sprayed/showered onto someone. Because, yes, we do have "wet_stacks", technically, but they're barely used so we might as well consider them inexistent. When used to dry someone up, it becomes "wet" and gains water equivalent to the amount of "wetness" on that person (which is calculated using negative values for the `fire_stacks` variable).

Towels can be used to dry up liquid puddles on the floor! That also makes them wet, and it's a little wasteful, much like most other ways to dry up puddles.

Wet towels become slightly grayer, to signify that they're wet.

A wet towel can be washed using a sink to completely get rid of the reagents in it, fully soaking it in water in the process. It can also be wrung using ctrl-click, to create a puddle of liquids 10u at a time at your feet! However, that's not super efficient, and about half of the reagents are lost that way. Likewise, you can right-click on buckets to wring the liquids out into it, with a 25% loss ratio instead of 50%. Finally, a wet towel can also be washed in a washing machine, to come out clean and dry! It also gives the freshly washed mood buff, like any other piece of clothing.

Towels can be crafting using two pieces of cloth in-hand, using the stack recipes. They can also be obtained from towel bins, which when mapped in will contain 10 of them, and which are otherwise functionally identical to bedsheet bins in how you can interact with them.

I also made sure to make it as explicit as possible in the examine what you can do with the towel, so that there's no need to look into the code to figure it out. I even added screentips for it, too!

#### Todo:
- [ ] Make using a towel that's wet and containing water usable to put out someone that's on fire, drying it up in the process.
- [ ] Add some towel bins to the Twin Nexus Hotel.
- [ ] Make water make you wet, to give you an opportunity to use towels.
- [ ] Let towels also allow you to remove the "something flammable" that's sometimes covering people, too.

Big thanks to @Ebin-Halcyon for the sprites!
![image](https://user-images.githubusercontent.com/58045821/209885604-3fddd7df-4a93-4b15-a259-9a750e824658.png)

In the future, provided we get sprites for it, we could add some towel holders that could be placed on walls, and could hold a towel, to allow it to dry or something like that, for some true bathroom experience.

I also fixed a few bugs along the way, I'm going to open a PR upstream to fix the mood buff for washed clothes, though, as that's an issue there too.

## How This Contributes To The Skyrat Roleplay Experience
Towels are great, I'm sure there's MANY people that have been feeling the need to use towels in-game after swimming in a pool or when in a sauna, and just never were able to.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>

Very nice indeed.  
![image](https://user-images.githubusercontent.com/58045821/209885727-518e7de0-9fed-42f8-bbf0-ce05082759e4.png)

</details>

## Changelog

:cl: GoldenAlpharex, with Halcyon for the amazing sprites
add: Towels! Wear them on your chest, your waist or your head, use them to dry up those that are wet or to dry up liquid puddles.
add: You can wash towels in sinks, or wring them, to remove the reagents from them, but the best way to have clean towels will always be a washing machine! Wringing towels leaves a mess behind, unless you do it in a bucket, which will make it transfer a portion of the liquids into it, losing some on the way.
fix: Fixed ears being displayed at all times if you're wearing something on your head, but not on your mask slot.
fix: Fixed the mood buff for when you wash clothes.
code: Improved some of the code related to interactions between mops/reagent_containers and liquid turfs, hopefully fixing some bugs on the way too!
/:cl:

Zonespace27 commented on code/datums/components/onwear_mood.dm:
```diff
@@ -7,7 +7,7 @@
 	var/slot_equip
 
-/datum/component/onwear_mood/Initialize(datum/mood_event/saved_event, examine_string, slot_equip = ITEM_SLOT_ON_BODY)
+/datum/component/onwear_mood/Initialize(datum/mood_event/saved_event_type, examine_string, slot_equip = ITEM_SLOT_ON_BODY)
```
undocced change

GoldenAlpharex replied:
Yeah, that's the one I'm going to bring up on /tg/ likely either later today or tomorrow, hence why I didn't document it.

Zonespace27 commented on code/modules/reagents/reagent_containers/cups/_cup.dm:
```diff
@@ -391,7 +389,7 @@
 				to_chat(user, "<span class='warning'>[src] is full!</span>")
 				return
-			O.reagents.remove_any(O.reagents.total_volume*SQUEEZING_DISPERSAL_PERCENT)
+			O.reagents.remove_any(O.reagents.total_volume * SQUEEZING_DISPERSAL_RATIO)
```
also undocced

GoldenAlpharex replied:
This is Skyrat stuff already, if you look up.

Zonespace27 replied:
Looks like I didn't look high enough then

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+		return TOOL_ACT_TOOLTYPE_SUCCESS
+
+	if(tool.use_tool(src, user, 0.5 SECONDS, volume = 50))
+		to_chat(user, span_notice("You disassemble the [src]."))
```
```suggestion
		to_chat(user, span_notice("You disassemble [src]."))
```

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+		return FALSE
+
+	if(amount)
+		to_chat(user, span_warning("The [src] must be empty first!"))
```
```suggestion
		to_chat(user, span_warning("[src] must be empty first!"))
```

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+	. = ..()
+	if(amount < 1)
+		. += "There are no towels in the bin."
+	else if(amount == 1)
```
27-28 strikes me as unneeded, you can just ternerize the towel/towels

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+
+	var/obj/item/towel/towel
+
+	if(towels.len > 0)
```
```suggestion
	if(length(towels))
```

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+ * * tk - Is the user trying to do this using telekinesis? Defaults to `FALSE`.
+ */
+/obj/structure/towel_bin/proc/take_towel_out(mob/user, tk = FALSE)
+	if(amount < 1)
```
(this and elsewhere)
the "if there's less than 1" syntax is a little strange

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+	var/obj/item/towel/towel
+
+	if(towels.len > 0)
+		towel = towels[towels.len]
```
```suggestion
		towel = towels[length(towels)]
```

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+	/// How many towels there is in the bin (separate from the towels list because we won't instanciate 10 towels per bin in existance).
+	var/amount = 10
+	/// The list of already-instanciated towels, for when people put them back in it.
+	var/list/towels = list()
```
can this be a lazylist?

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+
+	else if(amount && !hidden && attacking_item.w_class < WEIGHT_CLASS_BULKY) //make sure there's sheets to hide it among, make sure nothing else is hidden in there.
+		if(!user.transferItemToLoc(attacking_item, src))
+			to_chat(user, span_warning("\The [attacking_item] is stuck to your hand, you cannot hide it among the sheets!"))
```
\the is implicit
```suggestion
			to_chat(user, span_warning("[attacking_item] is stuck to your hand, you cannot hide it among the sheets!"))
```

Zonespace27 commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,148 @@
+		towels.Remove(towel)
+
+	else
+		towel = new /obj/item/towel(loc)
```
```suggestion
		towel = new (loc)
```

Zonespace27 commented on modular_skyrat/modules/liquids/code/reagent_containers.dm:
```diff
@@ -4,37 +4,43 @@
-	tempr.trans_to(my_beaker.reagents, tempr.total_volume)
-	to_chat(user, span_notice("You scoop up around [my_beaker.amount_per_transfer_from_this] units of liquids with [my_beaker]."))
+	tempr.trans_to(reagents, tempr.total_volume)
+	to_chat(user, span_notice("You scoop up around [amount_per_transfer_from_this] units of liquids with \the [src]."))
```
```suggestion
	to_chat(user, span_notice("You scoop up around [amount_per_transfer_from_this] units of liquids with [src]."))
```

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+	register_item_context()
+
+
+/obj/item/towel/examine(mob/user)
```
\the is implicit in all of these

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+/obj/item/towel/AltClick(mob/user)
+	. = ..()
+
+	if(. == FALSE)
```
why not just !. instead

GoldenAlpharex replied:
Because it can return `null` too, which isn't what I want to catch in this case. I explicitly want to catch only `FALSE`.

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+/obj/item/towel/CtrlClick(mob/user)
+	. = ..()
+
+	if(. == FALSE)
```
ditto

GoldenAlpharex replied:
Same answer for here too.

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+			to_chat(user, span_notice("You fold [src] up neatly."))
+			return
+
+		// No cyborgs past this point.
```
comment isn't quite accurate

GoldenAlpharex replied:
Now it will be, I messed up my first if() logic initially, whoops. But yeah, cyborgs with wet towels are caught prior to this point, and then not wet towels are caught here, so it's cyborg-proof past that.

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+
+	wet = new_wetness
+
+	color = wet ? "#CCCCCC" : null
```
maybe use `add_atom_color` instead?

GoldenAlpharex replied:
I don't think that's really making it better, I checked the code and it's mostly for when we plan on having multiple colors at once which don't appear at the same time, but rather based on priority, which isn't the case here.

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+		return
+
+	shape = new_shape
+	var/new_icon_state = "[base_icon_state][shape ? "-[shape]" : ""]"
```
why can't this be on 372

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+	set_wet(reagents.total_volume)
+	make_used(user, silent = TRUE)
+
+	to_chat(user, span_notice("You soak \the [src] with some liquids."))
```
implict \the

Zonespace27 commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,523 @@
+		to_chat(user, span_warning("Your [src] can't absorb any more liquid!"))
+		return TRUE
+
+	var/datum/reagents/tempr = liquids.take_reagents_flat(free_space)
```
what does tempr mean

OrionTheFox commented:
Bless, bless!!

Got a few things I wanna discuss and am glad to help with if you'd like:
\> Having the slot chosen by folding it feels tedious for the user; on oldbase it fit into all slots and changed the worn sprite based on which it was currently in, is there a non-shitcode way to recreate that? It felt much more intuitive to just place it in a slot and be done
\> Oldbase allowed them to be laid out on the ground by using it in-hand; this was actually really simple codewise iirc and just dropped the item similar to a bedsheet, but changed the icon to be laid out flat. Think that's possible too? Beaches neeeeed it...
\> The towel basket is tg's bedsheet basket, correct? You can place either in? Or are they unique items? (Im on mobile at work I didnt actually read the code if you hadnt guessed)
\> Gags? Gags? :eyes:

GoldenAlpharex commented:
> Got a few things I wanna discuss and am glad to help with if you'd like:
> Having the slot chosen by folding it feels tedious for the user; on oldbase it fit into all slots and changed the worn sprite based on which it was currently in, is there a non-shitcode way to recreate that? It felt much more intuitive to just place it in a slot and be done

I could probably figure out a way to do that, but honestly it's not quite like you're going to put it on and take it off constantly, I reckon this should work just fine for now. Could be a good idea for a future improvement, possibly.

> Oldbase allowed them to be laid out on the ground by using it in-hand; this was actually really simple codewise iirc and just dropped the item similar to a bedsheet, but changed the icon to be laid out flat. Think that's possible too? Beaches neeeeed it... 

I don't have a sprite for that, but yeah otherwise it would be relatively easy to do.

> The towel basket is tg's bedsheet basket, correct? You can place either in? Or are they unique items? (Im on mobile at work I didnt actually read the code if you hadnt guessed)

They're separate items, to keep the behavior that it had beforehand. Otherwise it would get a lot messier than it currently is, and would require a lot more code to support it.

> Gags? Gags? 👀

There's no GAGS support for it, but it's also the kind of thing where just applying a color would work. But, I mean, making it GAGS would probably be rather trivial, I just haven't done it myself because I didn't really feel like it was necessary right away.

dawsonkeyes commented:
why would you want a gags towel

OrionTheFox commented:
Honwstly I entirely forgot that we can set the color without gags so i call recolorable gags

And bc towels arent just white of course
I need my beach towels 4 roleplay.,..,

OrionTheFox commented on modular_skyrat/master_files/code/game/objects/structures/towel_bins.dm:
```diff
@@ -0,0 +1,146 @@
+ */
+/obj/structure/towel_bin/proc/take_towel_out(mob/user, tk = FALSE)
+	if(amount <= 0)
+		to_chat(user, span_warning("You can't figure out how to take a towel out of [src] when it doesn't contain any!"))
```
```suggestion
		balloon_alert(user, "bin is empty!")
```
All the homies hate to_chat failiures
Bitches LOVE balloon alerts

GoldenAlpharex replied:
This is just code that was copied out of the bedsheet bin code, I think.

OrionTheFox replied:
😭 another thing to fix on TG...... nooo....

OrionTheFox commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,522 @@
+
+	var/free_space = reagents.maximum_volume - reagents.total_volume
+	if(free_space <= 0)
+		to_chat(user, span_warning("Your [src] can't absorb any more liquid!"))
```
It would be nice to see lots of these attack to_chats become balloon_alerts but my brain doesnt work right now so I can't suggest good ideas, and I can't blame you if you don't want to or if you have the same issue

I just hate to_chats... grr....

OrionTheFox commented on modular_skyrat/master_files/code/modules/clothing/towels.dm:
```diff
@@ -0,0 +1,522 @@
+
+	if(ishuman(user) || iscyborg(user))
+		if(iscyborg(user) && wet) // Cyborgs can't wring towels.
+			to_chat(user, span_warning("Folding a wet towel doesn't really make sense. You stop yourself before doing that."))
```
```suggestion
			balloon_alert(user, "dry it out first!")
```
Balloon alert for failure is preferable 🥺
